crates/cervo-runtime/src/runtime.rs

```diff
@@ -98,10 +156,72 @@ impl Runtime {
         Ok(result)
     }
 
+    pub fn run_for_threaded(
+        &mut self,
+        duration: Duration,
+    ) -> Result<HashMap<BrainId, HashMap<AgentId, Response<'_>>>, CervoError> {
+        let start = Instant::now();
+        let mut any_executed = false;
+
+        let mut sorted_queue: Vec<Ticket> = Vec::with_capacity(self.queue.len());
+        while !self.queue.is_empty() {
+            sorted_queue.push(self.queue.pop().unwrap());
+        }
+
+        let queue = sorted_queue
+            .iter()
+            .filter_map(|ticket| {
+                if let Some(model) = self.models.iter().find(|m| m.id == ticket.1) {
+                    if model.needs_to_execute() || !any_executed {
+                        any_executed = true;
+                        return Some((ticket, model));
+                    }
+                }
+                None
+            })
+            .collect::<Vec<(&Ticket, &ModelState)>>();
+
+        let results = queue
+            .into_par_iter()
+            .map(|(ticket, model)| {
+                if start.elapsed() > duration {
+                    return None;
+                }
+                let time_remaining = duration.clone().saturating_sub(start.elapsed());
+                if model.can_run_in_time(time_remaining) {
```

Comment:
I don't think this check works as intended. Since you're cloning the duration above, you'll only ever check if the current iteration is < the max duration. That doesn't seem right.

Reply:
I think as a starting point we can use the following heuristic: Running for 1 second on 1 thread is the same as running for 0.25 seconds on four threads. So you can say `let duration = duration * rayon::current_num_threads();`. Then assume all jobs will take exactly as long as they say.  Then you can move this check into the above filter step without caring about threading.

Reply:
Given that `start` only represents the timestamp at which multithreading started, and the initial duration doesn't change, but only gets cloned to be accessed from inside multithreading, it does exit when intended. Not sure about how much overhead that causes though. 

With the above suggestion, would that not require a new function estimating the time it takes to subtract from the `duration * num_threads` before multithreading?

Reply:
Ooh, right. I misread the logic. Sorry. This isn't accumulating, it just computes how much time is left. That's fine! I do think it might still be easier to avoid branching in the multi-threaded code as we'll otherwise generate a lot of jobs do nothing and just churn the dispatcher.

I'm not sure I understand your question, but maybe I explained myself poorly. If we pretend the full time is actually the number of threads running in sequence, we can do something like this instead:

```rust
let mut available_cpu_time= duration * rayon::current_num_threads();
let mut selected_jobs = vec![];
let mut unselected_jobs = vec![];

while let Some(ticket) = self.queue.pop(} {
    let Some(model) = self.models.get(&ticket.1) else { continue; };

    if model.needs_to_run() && (selected_jobs.is_empty() || model.can_run_in(availabe_cpu_time) {
        available_cpu_time = available_cpu_time.saturating_sub(model.estimated_time());
        selected_jobs.push(ticket);
    } else {
       unselected_jobs.push(ticket);
    }
}

let results = selected_jobs.into_par_iter().map(|ticket| {
    let model = self.models.get(&ticket.1).unwrap();
    (ticket.1, model.run())
}).collect::<Vec<(_, _)>(); // collect necessary to conserve ordering

self.queue
    .extend(unselected_jobs.into_iter().map(|t| t.1).chain(results.iter().map(|(b, _)| *b))
    .map(|brain| {
        let gen = self.ticket_generation;
        self.ticket_generation += 1;
        Ticket(gen, b)
    });

// transpose Iter<(B, Res<V, E>)> into Iter<Res<(B, Val)>> before collecting
results
    .into_iter()
    .map(|(b, res)| res.map(|val| (b, val))
    .collect::<Result<_>>()
```

Reply:
(just off the top of my head, haven't tested implementing the above!)

Reply:
That seems good, good point. Will check that out in detail tomorrow!